File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.java
class Solution {
    public void dfs(
        List<String> result,
        Map<Character, List<Character>> dic, 
        String digits,
        int index,
        StringBuilder path
    ) {

        if(path.length() == digits.length()) {
            result.add(String.valueOf(path));
            return;
        }

        for(Character c : dic.get(digits.charAt(index))) {
            StringBuilder newPath = new StringBuilder(path).append(c);
            dfs(result, dic, digits, index + 1, newPath);
        }
    }
    
    public List<String> letterCombinations(String digits) {
        List<String> result = new ArrayList<>();
        
        // 예외 처리
        if(digits.length() == 0) {
            return result;
        }
        
        // 번호에 해당하는 문자 목록 구성
        Map<Character, List<Character>> dic = Map.ofEntries(
            Map.entry('0', List.of()), 
            Map.entry('1', List.of()), 
            Map.entry('2', List.of('a', 'b', 'c')), 
            Map.entry('3', List.of('d', 'e', 'f')), 
            Map.entry('4', List.of('g', 'h', 'i')), 
            Map.entry('5', List.of('j', 'k', 'l')), 
            Map.entry('6', List.of('m', 'n', 'o')), 
            Map.entry('7', List.of('p', 'q', 'r', 's')), 
            Map.entry('8', List.of('t', 'u', 'v')), 
            Map.entry('9', List.of('w', 'x', 'y', 'z'))
        );
        
        // 재귀 호출
        dfs(result, dic, digits, 0, new StringBuilder());
        return result;
    }
}

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.kt
class Solution {
    fun letterCombinations(digits: String): List<String> {
        // 결과를 저장할 리스트 선언
        val result: MutableList<String> = mutableListOf()

        // 예외 처리
        if(digits.isEmpty()) {
            return result
        }

        // 번호로 조합 가능한 문자 목록 구성
        val dic: Map<Char, List<Char>> = mapOf(
            '0' to listOf(),
            '1' to listOf(),
            '2' to listOf('a', 'b', 'c'),
            '3' to listOf('d', 'e', 'f'),
            '4' to listOf('g', 'h', 'i'),
            '5' to listOf('j', 'k', 'l'),
            '6' to listOf('m', 'n', 'o'),
            '7' to listOf('p', 'q', 'r', 's'),
            '8' to listOf('t', 'u', 'v'),
            '9' to listOf('w', 'x', 'y', 'z')
        )


        fun dfs(index: Int, path: StringBuilder) {
            // 끝까지 탐색했으면 결과를 저장하고 리턴
            if(path.length == digits.length) {
                result.add(path.toString())
                return
            }
            
            // 현재 자리 숫자에 대항하는 모든 문자열 탐색
            for(c in dic[digits[index]]!!) {
                // 현재자리 +1, 지금까지 구성된 문자열 path 이용하여 재귀 DFS
                dfs(index +1, StringBuilder(path).append(c))
            }
	    }
	
    // 현재자리 0, 빈 문자열 이용 DFS 시작
	dfs(0, StringBuilder())
	return result
    }
}